File: js/src/vector/traits.ts
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import { Vector } from './vector';
import { BoolVector } from './numeric';
import { fb, FieldBuilder, FieldNodeBuilder } from '../format/arrow';

export type Field = ( fb.Schema.Field | FieldBuilder );
export type FieldNode = ( fb.Message.FieldNode | FieldNodeBuilder );

function isField(x: any): x is Field {
    return x instanceof fb.Schema.Field || x instanceof FieldBuilder;
}

function isFieldNode(x: any): x is FieldNode {
    return x instanceof fb.Message.FieldNode || x instanceof FieldNodeBuilder;
}

export function isFieldArgv(x: any): x is { field: Field, fieldNode: FieldNode } {
    return x && isField(x.field) && isFieldNode(x.fieldNode);
}

export function isNullableArgv(x: any): x is { validity: Uint8Array } {
    return x && x.validity && ArrayBuffer.isView(x.validity) && x.validity instanceof Uint8Array;
}

type Ctor<TArgv> = new (argv: TArgv) => Vector;

export const nullableMixin = <T extends Vector, TArgv>(superclass: new (argv: TArgv) => T) =>
    class extends (superclass as Ctor<TArgv>) {
        readonly validity: Vector<boolean>;
        constructor(argv: TArgv & { validity: Uint8Array }) {
            super(argv);
            this.validity = new BoolVector({ data: argv.validity });
        }
        get(index: number) {
            return this.validity.get(index) ? super.get(index) : null;
        }
    };

export const fieldMixin = <T extends Vector, TArgv>(superclass: new (argv: TArgv) => T) =>
    class extends (superclass as Ctor<TArgv>) implements Vector {
        readonly field: Field;
        readonly type: string;
        readonly length: number;
        readonly stride: number;
        readonly nullable: boolean;
        readonly nullCount: number;
        readonly fieldNode: FieldNode;
        constructor(argv: TArgv & { field: Field, fieldNode: FieldNode }) {
            super(argv);
            const { field, fieldNode } = argv;
            this.field = field;
            this.fieldNode = fieldNode;
            this.nullable = field.nullable();
            this.type = fb.Schema.Type[field.typeType()];
            this.length = fieldNode.length().low | 0;
            this.nullCount = fieldNode.nullCount().low;
        }
        get name() { return this.field.name()!; }
        get metadata()  {
            const { field } = this, data = new Map<string, string>();
            for (let entry, key, i = -1, n = field && field.customMetadataLength() | 0; ++i < n;) {
                if ((entry = field.customMetadata(i)) && (key = entry.key()) != null) {
                    data.set(key, entry.value()!);
                }
            }
            return data;
        }
    };

File: matlab/src/matlab/+arrow/+type/+traits/traits.m
% Licensed to the Apache Software Foundation (ASF) under one or more
% contributor license agreements.  See the NOTICE file distributed with
% this work for additional information regarding copyright ownership.
% The ASF licenses this file to you under the Apache License, Version
% 2.0 (the "License"); you may not use this file except in compliance
% with the License.  You may obtain a copy of the License at
%
%   http://www.apache.org/licenses/LICENSE-2.0
%
% Unless required by applicable law or agreed to in writing, software
% distributed under the License is distributed on an "AS IS" BASIS,
% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
% implied.  See the License for the specific language governing
% permissions and limitations under the License.

function typeTraits = traits(type)
    % "Gateway" function that links an arrow Type ID enumeration (e.g.
    % arrow.type.ID.String) or a MATLAB class string (e.g. "datetime")
    % to associated type information.
    import arrow.type.traits.*
    import arrow.type.*
    
    if isa(type, "arrow.type.ID")
        switch type
            case ID.UInt8
                typeTraits = UInt8Traits();
            case ID.UInt16
                typeTraits = UInt16Traits();
            case ID.UInt32
                typeTraits = UInt32Traits();
            case ID.UInt64
                typeTraits = UInt64Traits();
            case ID.Int8
                typeTraits = Int8Traits();
            case ID.Int16
                typeTraits = Int16Traits();
            case ID.Int32
                typeTraits = Int32Traits();
            case ID.Int64
                typeTraits = Int64Traits();
            case ID.Float32
                typeTraits = Float32Traits();
            case ID.Float64
                typeTraits = Float64Traits();
            case ID.Boolean
                typeTraits = BooleanTraits();
            case ID.String
                typeTraits = StringTraits();
            case ID.Timestamp
                typeTraits = TimestampTraits();
            case ID.Time32
                typeTraits = Time32Traits();
            case ID.Time64
                typeTraits = Time64Traits();
            case ID.Date32
                typeTraits = Date32Traits();
            case ID.Date64
                typeTraits = Date64Traits();
            case ID.List
                typeTraits = ListTraits();
            case ID.Struct
                typeTraits = StructTraits();
            otherwise
                error("arrow:type:traits:UnsupportedArrowTypeID", "Unsupported Arrow type ID: " + type);
        end
    elseif isa(type, "string") % MATLAB class string
        switch type
            case "uint8"
                typeTraits = UInt8Traits();
            case "uint16"
                typeTraits = UInt16Traits();
            case "uint32"
                typeTraits = UInt32Traits();
            case "uint64"
                typeTraits = UInt64Traits();
            case "int8"
                typeTraits = Int8Traits();
            case "int16"
                typeTraits = Int16Traits();
            case "int32"
                typeTraits = Int32Traits();
            case "int64"
                typeTraits = Int64Traits();
            case "single"
                typeTraits = Float32Traits();
            case "double"
                typeTraits = Float64Traits();
            case "logical"
                typeTraits = BooleanTraits();
            case "string"
                typeTraits = StringTraits();
            case "datetime"
                typeTraits = TimestampTraits();
            case "duration"
                typeTraits = Time64Traits();
            case "table"
                typeTraits = StructTraits();
            otherwise
                error("arrow:type:traits:UnsupportedMatlabClass", "Unsupported MATLAB class: " + type);
        end
    else
        error("arrow:type:traits:UnsupportedInputType", "The input argument to the traits function " + ...
                                                        "must be a MATLAB class string or an arrow.type.ID enumeration.");
    end
end
